com.unity.cinemachine/Editor/Upgrader/CinemachineUpgradeManager.refactor.cs

```diff
@@ -0,0 +1,264 @@
+            }
+            m_CurrentSceneOrPrefab = string.Empty;
+        }
+
+        /// <summary>
+        /// For each scene:
+        /// - Upgrade referencable prefab instances without deleting obsolete components,
+        /// - Sync referencable prefab instances with their linked copies (to regain lost prefab instance modifications)
+        /// - Delete referencable prefab instance copies
+        /// - Update timeline references
+        /// </summary>
+        /// <param name="conversionLinksPerScene">Key: scene index, Value: List of conversion links</param>
+        void UpgradeReferencablePrefabInstances(Dictionary<int, List<ConversionLink>> conversionLinksPerScene)
+        {
+            for (var s = 0; s < m_SceneManager.SceneCount; ++s)
+            {
+                var scene = OpenScene(s);
+                var timelineManager = new TimelineManager(scene);
+                var upgradedObjects = new HashSet<GameObject>();
+                UpgradePrefabInstances(upgradedObjects, conversionLinksPerScene[s], timelineManager, true);
+                
+                EditorSceneManager.SaveScene(scene);
+            }
+
+            m_CurrentSceneOrPrefab = string.Empty;
+        }
+        
+        /// <summary>
+        /// For each scene:
+        /// - Upgrade non-referencable prefab instances without deleting obsolete components, sync with their
+        /// linked copies (to regain lost prefab instance modifications), delete copies, and update timeline references
+        /// - Upgrade non-prefabs, and update timeline references
+        /// - Fix animation references
+        /// - Fix object references
+        /// - Delete obsolete components
+        /// - Restore timeline names to their original names
+        /// </summary>
+        /// <param name="conversionLinksPerScene">Key: scene index, Value: List of conversion links</param>
+        /// <param name="timelineRenames">Timeline rename mapping</param>
+        void UpgradeRemaining(
+            Dictionary<int, List<ConversionLink>> conversionLinksPerScene,
+            Dictionary<string, string> timelineRenames)
+        {
+            for (var s = 0; s < m_SceneManager.SceneCount; ++s)
+            {
+                var scene = OpenScene(s);
+                var timelineManager = new TimelineManager(scene);
+                var upgradedObjects = new HashSet<GameObject>();
+                
+                UpgradePrefabInstances(upgradedObjects, conversionLinksPerScene[s], timelineManager, false);
+                
+                var rootObjects = scene.GetRootGameObjects();
+                var upgradables = GetUpgradables(rootObjects, m_ObjectUpgrader.RootUpgradeComponentTypes, true);
+                UpgradeNonPrefabs(upgradables, upgradedObjects, timelineManager);
+                
+                // restore dolly references in prefab instances
+                foreach (var go in upgradables) 
+                    UpgradeObjectComponents(go, null);
```

Comment:
I don't understand why this is here.  Didn't you just do this in UpgradeNonPrefabs()?

Reply:
oh I see... this is re-upgrading the prefab instances that were skipped by UpgradeNonPrefabs.  Didn't we upgrade those earlier?  What case is this handling?

Reply:
This restores any un-restored references to CmPath/Splines in prefab instances. Because we kept obsolete components up to this point, we can use those to correctly link SplineDolly's and SplineCarts to their splines.

This step could be avoided by only upgrading referencables (prefab asset, prefab instance, non prefab), and then non-referencables (prefab asset, prefab instance, non prefab). Right now, it is not exactly like that, so this extra pass at the end is needed.

Reply:
So, what specific case is handled by this?  Is it a prefab instances with an override that points to a non-prefab path?

Reply:
Yes, that for example.

Reply:
Here is what I don't understand.  When the prefab instance got upgraded, the spline will have been set to null because the non-prefab path had not been updated yet, so the reference to the old path would have been lost.  Why wasn't it?

Reply:
got it, thanks